2021.04+imx fixes for apalis-imx6, apalis-imx8, and imx6ull-evk

MrCry0 opened the pull request:
Fixes for:
- building apalis-imx6[-sec];
- building apalis-imx8;
- starting SPL on apalis-imx6;
- starting OP-TEE and u-boot on imx6ull-evk.

- the latest backports from 2020.04

mike-scott committed:
[FIO extras] board: imx6ullevk: for OP-TEE lower ram_size by CONFIG_OPTEE_TZDRAM_SIZE

Signed-off-by: Michael Scott <mike@foundries.io>
Signed-off-by: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>

mrnuke committed:
[FIO fromtree] Revert "spl: Drop bd_info in the data section"

This reverts commit 38d6b7ebdaee3e0e8426ef1b9df88bdce8ae2e75.

struct global_data contains a pointer to the bd_info structure. This
pointer was populated spl_set_bd() to a pre-allocated bd_info in the
".data" section. The referenced commit replaced this mechanism to one
that uses malloc(). That new mechanism is only used if SPL_ALLOC_BD=y.
which very few boards do.

The result is that (struct global_data)->bd is NULL in SPL on most
platforms. This breaks falcon mode, since arch_fixup_fdt() tries to
access (struct global_data)->bd and set the "/memory" node in the
devicetree. The result is that the "/memory" node contains garbage
values, causing linux to panic() as it sets up the page table.

Instead of trying to fix the mess, potentially causing other issues,
revert to the code that worked, while this change is reworked.

Signed-off-by: Alexandru Gagniuc <mr.nuke.me@gmail.com>
Signed-off-by: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>

MrCry0 committed:
[FIO toup-squash] fit_image: use sha256 algo for images hash signing

In case the STRICT mode of FIT signature checking enabled, crc32,
sha1, and md5 algorithms are disabled.
Use sha256 for signing images. This fixes u-boot.img building
error [1].

[1]
MKIMAGE u-boot.img
Unsupported hash algorithm (crc32) for 'hash-1' hash node in 'firmware-1' image node
./tools/mkimage Can't add hashes to FIT blob: -93
make: *** [Makefile:1381: u-boot.img] Error 255

Fixes: 4b14e827d5 ("[FIO extras] fit-image: dont use CRC32, MD5 and SHA1 if SIGNATURE_STRICT")
Signed-off-by: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>

Tim-Anderson reviewed (approved):
LGTM

ricardosalveti commented on common/image-fit.c:
```diff
@@ -1224,6 +1224,9 @@ int calculate_hash(const void *data, int data_len, const char *algo,
 			     (unsigned char *)value, CHUNKSZ_SHA1);
 		*value_len = 20;
+	} else if (IMAGE_ENABLE_MD5 && strcmp(algo, "md5") == 0) {
+		md5_wd((unsigned char *)data, data_len, value, CHUNKSZ_MD5);
```
Is this function defined when MD5 is not enabled/used?

MrCry0 replied:
@ricardosalveti The same cases as for crc32/sha1. Look at lines 1216-1217.

ricardosalveti replied:
Indeed, but curious why we had this block protected before and why that is not needed anymore.

MrCry0 replied:
It was just protected with 2 different ifdef's because these algos crc32, sha1, and md5 were just non-contiguous :)

So this patch doesn't change logic at all, it just allows to remove extra ifdef/endif lines.

ldts replied:
wont this patch increase the size of the binary? wasnt the reason for the patch to make sure we save ~4k

MrCry0 replied:
@ldts No.

ldts replied:
No what?

MrCry0 replied:
No, this patch won't increase the size of the binary.

MrCry0 replied:
@ldts This particular patch makes cosmetic changes only, it doesn't change the logic of removing this code by a preprocessor. Does it solve your question?

ldts commented on include/spl.h:
```diff
@@ -694,7 +694,7 @@ struct bl_params *bl2_plat_get_bl31_params_v2_default(uintptr_t bl32_entry,
  * @arg3: non-secure entry address (ARMv7 bootarg #0)
  */
-void spl_optee_entry(void *arg0, void *arg1, void *arg2, void *arg3);
+void __noreturn spl_optee_entry(void *arg0, void *arg1, void *arg2, void *arg3);
```
why is it not an issue upstream?

MrCry0 replied:
It is. That's why it is fixed in the 2nd version of the patchset for upstream:
https://patchwork.ozlabs.org/project/uboot/list/?series=268058

MrCry0 replied:
@ldts Have I solved your question?

igoropaniuk committed:
[FIO internal] imx: secondary_boot: enable support for mx8

Enable support of imx_secondary_boot and imx_is_closed cmds
for MX8 SoCs (imx8qm and imx8qxp).

Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>

igoropaniuk committed:
[FIO internal] imx: runtime image offset detection for imx8

Add support of next image ram offset detection (in SPL) for iMX8QM/iMX8QXP.

Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>

igoropaniuk committed:
[FIO internal] fastboot: fb_fsl: handle secondary boot images for imx8

From 5.8.2.2.1 High Level eMMC Boot Flow Note, for the eMMC boot scenarios
where the images are located in the boot partition the boot image set
selection is done based on BOOT_PARTITION_ENABLE eMMC ECSD register values,
which means that secondary boot image set should be flashed to boot1
partition to the same offset the primary one.

Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>

igoropaniuk committed:
[FIO internal] imx: detect boot0/boot1 partition to boot in runtime

As secondary boot image set should be flashed to boot1
partition to the same offset the primary one, we should also detect in
runtime what boot partition to use for reading FIT image.

Signed-off-by: Igor Opaniuk <igor.opaniuk@foundries.io>

MrCry0 committed:
[FIO toup] mx6ullevk: Set CONFIG_SYS_BOOTMAPSZ to the memory for relocation

This patch sets CONFIG_SYS_BOOTMAPSZ to the amount of memory
available to safely contain a kernel, device tree and initrd for
relocation. The way with setting fdt_high as 0xffffffff to disable
device tree relocation is removed.

Signed-off-by: Oleksandr Suvorov <oleksandr.suvorov@foundries.io>

mike-scott committed:
[FIO toup] apalis-imx8: allow dropping CONFIG_CMD_USB requirement

When CONFIG_DISTRO_DEFAULTS is not defined, don't include
config_distro_bootcmd.h. This will break the build based
on the stock defines:
        func(MMC, mmc, 1) \
        func(MMC, mmc, 2) \
        func(MMC, mmc, 0) \
        func(USB, usb, 0)

Signed-off-by: Michael Scott <mike@foundries.io>

vanmaegima reviewed (approved):
LGTM
